### PR TITLE
SA-2541: Get-JCResults Parallel Implementation

### PR DESCRIPTION
**Pull request by gweinjc**
## Issues
* [SA-2541](https://jumpcloud.atlassian.net/browse/SA-2541) - Get-JCResults Parallel Implementation

## What does this solve?
This is laying the groundwork for allowing customers to utilize PowerShell 7 parallel processing when using specific Get functions. The Get-JCResults private function is intended to replace existing while loops to make API calls and return the results. A new boolean parameter, `-parallel $true/$false`, was introduced to determine if the API calls will be made in parallel or sequential. 

The next step is going to be removing the while loops in favor of this new function and introduce a `-parallel $true/$false` parameter in applicable functions

## Is there anything particularly tricky?
Parallel processing can only be utilized on PowerShell versions 7+. There are checks in place to determine if the current session has a compatible version as well as the `-parallel` parameter set to `true`.

if the version is not supported and `-parallel $true`, a warning will be displayed informing the user of the compatibility issue and the calls will be made sequentially

## How should this be tested?
Test to verify that both $rawResults contain the same information
```powershell
Import-Module "/Users/USERNAME/Documents/GitHub/support/PowerShell/JumpCloud Module/JumpCloud.psd1" -force
Import-Module "/Users/USERNAME/Documents/GitHub/support/PowerShell/JumpCloud Module/Private/NestedFunctions/Get-JCUserAgent.ps1" -force
Import-Module "/Users/USERNAME/Documents/GitHub/support/PowerShell/JumpCloud Module/Private/NestedFunctions/Get-JCResults.ps1" -force
Connect-JCOnline APIKEY
$JCUrlBasePath = 'https://console.jumpcloud.com'
$Group_ID = 'GROUPID'
$limitURL = "{0}/api/v2/usergroups/{1}/members" -f $JCUrlBasePath, $Group_ID
$rawResults = Get-JCResults -Url $limitURL -method "GET" -limit 100 -parallel $true -debug
$rawResults = Get-JCResults -Url $limitURL -method "GET" -limit 100 -parallel $false -debug
```
Measure time to complete for both functions (parallel *should* be faster, but not guaranteed) 
```powershell
Measure-Command { Get-JCResults -Url $limitURL -method "GET" -limit 100 -parallel $false }
Measure-Command { Get-JCResults -Url $limitURL -method "GET" -limit 100 -parallel $true }
```

### CODE REVIEW

**Comment by jworkmanjc** [comment]
Tested, on a 750 user org. My only thought is that instead of releasing to Main branch and releasing, let's just create a `JumpCloudModule_1.23.0` branch, merge this PR into that branch and when we actually enable this parallel results feature on endpoints let's release 1.23.0, I don't see a need to merge this into main branch just yet.


Screen Shot 2022-05-26 at 8 44 30 AM

File: PowerShell/JumpCloud Module/Private/NestedFunctions/Get-JCResults.ps1 [hunk 1]
```diff
@@ -5,7 +5,8 @@ Function Get-JCResults
         [Parameter(Mandatory = $true, HelpMessage = 'URL of Endpoint')][ValidateNotNullOrEmpty()]$URL,
         [Parameter(Mandatory = $true, HelpMessage = 'Method of WebRequest')][ValidateNotNullOrEmpty()]$method,
         [Parameter(Mandatory = $true, HelpMessage = 'Limit of WebRequest')][ValidateNotNullOrEmpty()]$limit,
-        [Parameter(Mandatory = $false, HelpMessage = 'Body of WebRequest, if required')]$body
+        [Parameter(Mandatory = $false, HelpMessage = 'Body of WebRequest, if required')]$body,
+        [Parameter(Mandatory = $false, HelpMessage = 'Boolean: True to run in parallel, False to run in sequential; Default value: false')][bool]$parallel = $false
```
Comment (jworkmanjc) [inline PowerShell/JumpCloud Module/Private/NestedFunctions/Get-JCResults.ps1:9]: Great, I think after we are done and confident we can switch this to default to parallel

File: PowerShell/JumpCloud Module/Private/NestedFunctions/Get-JCResults.ps1 [hunk 2]
```diff
@@ -17,36 +18,154 @@ Function Get-JCResults
...
+        catch {
+            throw $_
+        }
+
+        # Find total amount of results
         $totalCount = $response.Headers."x-total-count"
         $totalCount = [int]$totalCount.Trim()
         Write-Debug "total count: $totalCount"
+
+        # Divide amount of results by limit to find amount of passes to fully collect results
```
[the remark this answers is not in the document]
Comment (gweinjc) [reply PowerShell/JumpCloud Module/Private/NestedFunctions/Get-JCResults.ps1]: Done!

**Review by kmaranionjc — APPROVED** [review]
Nice work!

**Review by jworkmanjc — APPROVED** [review]
Thank you for the work on this ✅

**Pull request merged**